### PR TITLE
Adding concurrent request capabilities

### PR DESCRIPTION
**Pull request by alanvoss**
Per @manukall 's original request in https://github.com/PSPDFKit-labs/bypass/issues/12 .

We, too, were having problems at our organization with this.  It seems that keeping a running `List` of the calling `pid`s is sufficient.

I've included his test in the pull, and fixed it, I believe.  Commits squashed.

Closes PSPDFKit-labs/bypass#12 .

### CODE REVIEW

File: test/bypass_test.exs [hunk 1]
```diff
@@ -140,6 +140,22 @@ defmodule BypassTest do
...
+    end)
+    tasks = Enum.map(1..5, fn _ ->
+      Task.async(fn -> request(bypass.port) end)
+    end)
+    Enum.map(tasks, fn task ->
+      Task.await(task)
+      assert_receive :request_received
+    end)
+  end
+
```
Comment (alanvoss) [inline test/bypass_test.exs]: this is verbatim from

https://github.com/manukall/bypass/blob/concurrent_requests_test/test/bypass_test.exs#L118-L132

**Comment by colinrymer** [comment]
@MSch Any feedback on this PR? We'd really love to see this functionality supported and don't want to have to maintain our own fork.

File: lib/bypass/instance.ex [hunk 2]
```diff
@@ -81,7 +81,7 @@ defmodule Bypass.Instance do
   end
   defp do_handle_call(:down, from, %{socket: socket, ref: ref} = state) when not is_nil(socket) do
     nil = state.caller_awaiting_down  # assertion
-    if state.retained_plug != nil do
+    if state.retained_plugs != [] do
```
Comment (alco) [inline lib/bypass/instance.ex]: This waits for all retained plugs to finish instead of the specific plug associated with this Bypass instance.

Comment (alco) [inline lib/bypass/instance.ex]: Calling `down()` on multiple Bypass instances concurrently will trigger the assertion above. `caller_awaiting_down` should also be turned into a list and handled property in `dispatch_awaiting_caller()`.

Comment (alanvoss) [inline lib/bypass/instance.ex]: Thanks for taking a look.  I'll be sure to circle back to this ASAP.

Comment (alanvoss) [inline lib/bypass/instance.ex]: @alco  If we are doing concurrent requests, then "the specific plug associated with this Bypass instance" doesn't really hold -- there are multiple plugs associated with this instance's `state`.  We need to wait for all of them to finish.

 I think your second statement makes perfect sense to me, and definitely was something I neglected.

Can you please expand on your first statement and explain a little more?

Comment (alanvoss) [inline lib/bypass/instance.ex]: I am actively working on something that I think will supplant this, anyway, but I want to make sure I understand what you are saying before I continue.

**Comment by alanvoss** [comment]
I will be closing this PR soon in favor of another one.  Let's not merge this one.

**Comment by alanvoss** [comment]
Officially closing in favor of https://github.com/PSPDFKit-labs/bypass/pull/36 .

**Pull request closed without merging**